File: app/src/main/java/com/ravisharma/playbackmusic/Prefrences/PrefManager.java
package com.ravisharma.playbackmusic.Prefrences;

import android.content.Context;
import android.content.SharedPreferences;

import com.ravisharma.playbackmusic.R;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public class PrefManager {

    SharedPreferences sp;
    SharedPreferences.Editor ed;
    private String Playlist;

    public PrefManager(Context c) {
        sp = c.getSharedPreferences(c.getString(R.string.playback_info), Context.MODE_PRIVATE);
        ed = sp.edit();
        Playlist = c.getString(R.string.playLists);
    }

    public void storeInfo(String key, String data) {
        ed.putString(key, data);
        ed.commit();
    }

    public void storeInfo(String key, boolean data) {
        ed.putBoolean(key, data);
        ed.commit();
    }

    public String get_s_Info(String key) {
        return sp.getString(key, null);
    }

    public boolean get_b_Info(String key) {
        return sp.getBoolean(key, false);
    }

    public void createNewPlaylist(String playlistName) {
        Set<String> list = sp.getStringSet(Playlist, null);
        List<String> l;
        if (list == null) {
            l = new ArrayList<>();
        } else {
            l = new ArrayList<>(list);
        }
        l.add(playlistName);
        Set<String> list2 = new LinkedHashSet<>(l);
        ed.putStringSet(Playlist, list2);
        ed.commit();
    }

    public ArrayList<String> getAllPlaylist() {
        Set<String> list = sp.getStringSet(Playlist, null);
        if (list == null) {
            list = new LinkedHashSet<>();
        }
        return new ArrayList<>(list);
    }

    public void deletePlaylist(String playlistName) {
        Set<String> list = sp.getStringSet(Playlist, null);
        List<String> l = new ArrayList<>(list);
        l.remove(playlistName);

        Set<String> list2 = new LinkedHashSet<>(l);
        ed.putStringSet(Playlist, list2);
        ed.commit();
    }
}

File: app/src/main/java/com/ravisharma/playbackmusic/prefrences/PrefManager.kt
package com.ravisharma.playbackmusic.prefrences

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.SharedPreferencesMigration
import androidx.datastore.preferences.core.*
import androidx.datastore.preferences.preferencesDataStore
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.asLiveData
import com.ravisharma.playbackmusic.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import java.io.IOException

private val Context.dataStore by preferencesDataStore(
    name = "playback_info",
    produceMigrations = ::sharedPreferencesMigration
)

private fun sharedPreferencesMigration(context: Context) =
    listOf(SharedPreferencesMigration(context, context.getString(R.string.playback_info)))

class PrefManager(c: Context) {

    private val Playlist: String = c.getString(R.string.playLists)

    private val dataStore = c.dataStore

    fun clearAllData() {
        runBlocking {
            withContext(Dispatchers.Default) {
                dataStore.edit {
                    it.remove(intPreferencesKey("ID"))
                    it.remove(intPreferencesKey("position"))
                    it.remove(booleanPreferencesKey("Shuffle"))
                    it.remove(booleanPreferencesKey("RepeatOne"))
                    it.remove(booleanPreferencesKey("Repeat"))
                    it.remove(booleanPreferencesKey("Songs"))
                    it.remove(booleanPreferencesKey("Started"))
                }
            }
        }
    }

    fun storeInfo(key: String?, data: String) {
        runBlocking {
            withContext(Dispatchers.Default) {
                dataStore.edit {
                    if (data == "remove") {
                        it.remove(stringPreferencesKey(key!!))
                    } else {
                        it[stringPreferencesKey(key!!)] = data
                    }
                }
            }
        }
    }

    fun storeInfo(key: String?, data: Boolean) {
        runBlocking {
            withContext(Dispatchers.Default) {
                dataStore.edit {
                    it[booleanPreferencesKey(key!!)] = data
                }
            }
        }
    }

    fun storeAppVersion(data: Int) {
        runBlocking {
            withContext(Dispatchers.Default) {
                dataStore.edit {
                    it[intPreferencesKey("appVersion")] = data
                }
            }
        }
    }

    val appVersion: Int
        get() = runBlocking {
            withContext(Dispatchers.Default) {
                dataStore.getValueFlow(intPreferencesKey("appVersion"), -1).first()
            }
        }

    fun get_s_Info(key: String?): String {
        return runBlocking { dataStore.getValueFlow(stringPreferencesKey(key!!), "").first() }
    }

    fun get_b_Info(key: String?): Boolean {
        return runBlocking { dataStore.getValueFlow(booleanPreferencesKey(key!!), false).first() }
    }

    suspend fun createNewPlaylist(playlistName: String) {
        val list =
            dataStore.getValueFlow(stringSetPreferencesKey(Playlist), LinkedHashSet())
                .first()
        val l: MutableList<String> = if (list == null) {
            ArrayList()
        } else {
            ArrayList(list)
        }
        l.add(playlistName)
        val list2: Set<String> = LinkedHashSet(l)

        dataStore.edit {
            it[stringSetPreferencesKey(Playlist)] = list2
        }
    }

    suspend fun renamePlaylist(oldName: String, newName: String) {
        val list =
            dataStore.getValueFlow(stringSetPreferencesKey(Playlist), LinkedHashSet())
                .first()
        val l: MutableList<String>
        if (list != null) {

            l = ArrayList(list)
            l[l.indexOf(oldName)] = newName
            val list2: Set<String> = LinkedHashSet(l)

            dataStore.edit {
                it[stringSetPreferencesKey(Playlist)] = list2
            }
        }
    }

    fun fetchAllPlayList() =
        dataStore.getValueFlow(
            stringSetPreferencesKey(Playlist),
            LinkedHashSet()
        ).map {
            ArrayList(LinkedHashSet(it))
        }.asLiveData()

    suspend fun deletePlaylist(playlistName: String) {
        val list = dataStore.getValueFlow(
            stringSetPreferencesKey(Playlist),
            LinkedHashSet()
        ).first()

        val l: MutableList<String> = ArrayList(list)
        l.remove(playlistName)
        val list2: Set<String> = LinkedHashSet(l)

        dataStore.edit {
            it[stringSetPreferencesKey(Playlist)] = list2
        }
    }
}

fun <T> DataStore<Preferences>.getValueFlow(
    key: Preferences.Key<T>,
    defaultValue: T,
): Flow<T> {
    return this.data
        .catch { exception ->
            if (exception is IOException) {
                emit(emptyPreferences())
            } else {
                throw exception
            }
        }.map { preferences ->
            preferences[key] ?: defaultValue
        }
}